src/actions/debuggee.js

```diff
@@ -5,10 +5,13 @@
 // @flow
 
 import type { Action, ThunkArgs } from "./types";
+import { closeTabsForMissingThreads } from "./tabs";
 
 export function updateWorkers() {
-  return async function({ dispatch, client }: ThunkArgs) {
+  return async function({ dispatch, getState, client }: ThunkArgs) {
     const { workers } = await client.fetchWorkers();
     dispatch(({ type: "SET_WORKERS", workers }: Action));
+
+    dispatch(closeTabsForMissingThreads(workers));
```

Comment:
I think the cleanest thing to do would be to remove the sources for the removed workers. We could do this like so:

`dispatch(({ type: "SET_WORKERS", workers, destroyed: destroyedWorkers }: Action));`

in reducers/sources

```js
removeSourcesForWorkers(state, removedWorkers)
```

There might be other state that is now out-of-date because of the worker list has changed. For instance the currentThread might need to be reset too. Also, as a general rule of thumb, we try to keep our actions fairly high-level and not too implementation specific. It is better if they are more general-purpose like "newSources" or "updateWorkers"...

Reply:
I can try this, but I implemented the patch this way because there isn't a notion currently of being able to remove sources.  Adding this seems like it can affect quite a few places in the debugger, since sources are manipulated in so many places.  Allowing the sources to continue to exist, but removing them from the UI if they become obsolete (e.g. the worker they are associated with has closed) seemed a more robust design.

While closeTabsForMissingThreads is a pretty specific method, we do need to go through tab specific logic in order to get good behavior when removing sources --- if we just drop all tabs for removed sources then other parts of the editor pane can go out of sync (e.g. showing the contents of a tab that no longer exists).